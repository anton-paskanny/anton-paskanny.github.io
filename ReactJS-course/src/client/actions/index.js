export const SEARCH_MOVIE = 'SEARCH_MOVIE'

export const searchMovie = (type, value) => ({
    type: SEARCH_MOVIE,
    payload: {
        type,
        value
    }
});