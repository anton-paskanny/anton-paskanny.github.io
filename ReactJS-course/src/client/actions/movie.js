export const FETCH_MOVIE = 'FETCH_MOVIE';
export const FETCH_MOVIE_SUCCESS = 'FETCH_MOVIE_SUCCESS';
export const FETCH_MOVIE_ERROR = 'FETCH_MOVIE_ERROR';
export const SELECT_MOVIE = 'SELECT_MOVIE';
export const RESET_MOVIE = 'RESET_MOVIE';

export const fetchMovie = url => ({
    type: FETCH_MOVIE,
    url
});

export const fetchMovieSuccess = movie => ({
    type: FETCH_MOVIE_SUCCESS,
    movie
});

export const fetchMovieError = err => ({
    type: FETCH_MOVIE_ERROR,
    err
});

export const selectMovie = movie => ({
    type: SELECT_MOVIE,
    movie
});

export const resetMovie = () => ({
    type: RESET_MOVIE
});