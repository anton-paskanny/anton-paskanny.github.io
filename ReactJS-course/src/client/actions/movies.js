export const FETCH_MOVIES = 'FETCH_MOVIES';
export const FETCH_MOVIES_SUCCESS = 'FETCH_MOVIES_SUCCESS';
export const FETCH_MOVIES_BY_GENRES = 'FETCH_MOVIES_BY_GENRES';
export const FETCH_MOVIES_BY_GENRES_SUCCESS = 'FETCH_MOVIES_BY_GENRES_SUCCESS';
export const FETCH_MOVIES_ERROR = 'FETCH_MOVIES_ERROR';

export const fetchMovies = url => ({
    type: FETCH_MOVIES,
    url
});

export const fetchMoviesSuccess = movies => ({
    type: FETCH_MOVIES_SUCCESS,
    movies
});

export const fetchMoviesByGenres = url => ({
    type: FETCH_MOVIES_BY_GENRES,
    url
});

export const fetchMoviesByGenresSuccess = movies => ({
    type: FETCH_MOVIES_BY_GENRES_SUCCESS,
    movies
});

export const fetchMoviesError = err => ({
    type: FETCH_MOVIES_ERROR,
    err
});