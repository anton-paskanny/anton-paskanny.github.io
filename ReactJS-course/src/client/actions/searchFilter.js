export const TOGGLE_SEARCH = 'TOGGLE_SEARCH';

export const toggleSearch = type => ({
    type: TOGGLE_SEARCH,
    payload: type
});