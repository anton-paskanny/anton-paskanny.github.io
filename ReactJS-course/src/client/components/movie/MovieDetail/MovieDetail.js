import React, { Component } from 'react';

import Spinner from '../../shared/Spinner/Spinner';

import { URL_BASE } from '../../../utils';

import './styles.css';

class MovieDetail extends Component {
  componentDidMount() {
    this.getMovieData();
  }

  componentDidUpdate(prevProps) {
    if (prevProps.id !== this.props.id) {
      this.getMovieData();
    }
  }

  componentWillUnmount() {
    this.props.resetMovie();
  }

  renderGenres(genres) {
    return genres && genres.map(genre => (
      <li key={genre} className="movie-detail__genre">{genre}</li>
    ));
  }

  getMovieData() {
    const URL = `${URL_BASE}/${this.props.id}`;
    this.props.fetchMovie(URL);
  }

  render() {
    if (!this.props.selectedMovie || this.props.isFetching) return <Spinner />;

    const {
      title, overview, poster_path, genres, runtime, release_date, vote_average,
    } = this.props.selectedMovie;

    return (
      <div className="movie-detail">
          <figure className="movie-detail__img-placeholder">
            <img className="movie-detail__img" src={poster_path} alt={title} />
          </figure>
          <div className="movie-detail__info-wrapper">
            <div className="movie-detail__title-wrapper">
              <h2 className="movie-detail__title">{title}</h2>
              <p className="movie-detail__rating">
                  <span>{vote_average}</span>
              </p>
            </div>
            <ul className="movie-detail__genres">{this.renderGenres(genres)}</ul>
            <p className="movie-detail__wrapper">
              <time className="movie-detail__release-date" dateTime={release_date}>{release_date.slice(0, 4)} min</time>
              <span className="movie-detail__duration">{runtime}</span>
            </p>
            <p className="movie-detail__desc">{overview}</p>
          </div>
      </div>
    );
  }
}

export default MovieDetail;
