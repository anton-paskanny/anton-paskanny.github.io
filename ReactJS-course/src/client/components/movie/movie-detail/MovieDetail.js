import React, { Component } from 'react';

import styles from './styles.css';

class MovieDetail extends Component {
    constructor(props) {
        super(props);
        this.state = {
            isLoaded: false,
            movie: {}
        }
    }

    componentDidMount() {
        this.getMovieData(this.props.id);
    }

    componentDidUpdate(nextProps, nextState) {
        if (nextProps.id !== this.props.id) {
            this.getMovieData(nextProps.id);
        }  
    }

    renderGenres(genres) {
        return genres && genres.map(genre => (
            <li key={genre} className="movie-detail__genre">{genre}</li>
        ))
    }

    getMovieData(id) {
        const URL = `http://react-cdp-api.herokuapp.com/movies/${id}`;
        
        fetch(URL)
        .then(res => res.json())
        .then(data => this.setState({
            isLoaded: true,
            movie: data
        }));
    }

    render() {
        if (!this.state.isLoaded) return <p>Loading...</p>
        
        const { title, overview, poster_path, genres, runtime, release_date, vote_average } = this.state.movie;
        
        return (
            <div className="movie-detail">
                <figure className="movie-detail__img-placeholder">
                    <img className="movie-detail__img" src={poster_path} alt={title} />
                </figure>
                <div className="movie-detail__info-wrapper">
                    <div className="movie-detail__title-wrapper">
                        <h2 className="movie-detail__title">{title}</h2>
                        <p className="movie-detail__rating">
                            <span>{vote_average}</span>
                        </p>
                    </div>
                    <ul className="movie-detail__genres">{this.renderGenres(genres)}</ul>
                    <p className="movie-detail__wrapper">
                        <time className="movie-detail__release-date" dateTime={release_date}>{release_date.slice(0, 4)} min</time>
                        <span className="movie-detail__duration">{runtime}</span>
                    </p>
                    <p className="movie-detail__desc">{overview}</p>
                </div>
            </div>
        )
    }
}

export default MovieDetail;