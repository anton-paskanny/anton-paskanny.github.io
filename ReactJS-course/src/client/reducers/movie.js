import { loop, Cmd } from 'redux-loop';

import {
  FETCH_MOVIE,
  FETCH_MOVIE_SUCCESS,
  FETCH_MOVIE_ERROR,
  RESET_MOVIE,
  fetchMovieSuccess,
  fetchMovieError,
} from '../actions/movie';

const initialState = {
  isFetching: false,
  data: null,
  err: null,
};

const makeRequest = url => fetch(url).then(res => res.json());

const movieReducer = (state = initialState, action) => {
  switch (action.type) {
    case FETCH_MOVIE:
      return loop(
        { ...state, isFetching: true },
        Cmd.run(makeRequest, {
          successActionCreator: fetchMovieSuccess,
          failActionCreator: fetchMovieError,
          args: [action.url],
        }),
      );
    case FETCH_MOVIE_SUCCESS:
      return { ...state, data: action.movie, isFetching: false };
    case FETCH_MOVIE_ERROR:
      return { ...state, err: action.err, isFetching: false };
    case RESET_MOVIE:
      return { ...state, data: null };
    default:
      return state;
  }
};

export default movieReducer;
