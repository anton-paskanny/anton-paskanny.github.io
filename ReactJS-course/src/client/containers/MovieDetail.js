import { connect } from 'react-redux';
import { fetchMovie, resetMovie } from '../actions/movie';

import MovieDetail from '../components/Movie/MovieDetail/MovieDetail';

const mapStateToProps = state => ({
    selectedMovie: state.selectedMovie.data,
    isFetching: state.selectedMovie.isFetching
});

const mapDispatchToProps = {
    fetchMovie,
    resetMovie
}

export default connect(mapStateToProps, mapDispatchToProps)(MovieDetail);