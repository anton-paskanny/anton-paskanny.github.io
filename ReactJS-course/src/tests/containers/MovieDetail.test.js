import React from 'react';
import { Provider } from 'react-redux'
import { mount } from 'enzyme';
import configureMockStore from 'redux-mock-store';

import { resetMovie } from '../../client/actions/movie';

import MovieDetail from '../../client/containers/MovieDetail';


describe("MovieDetail container", () => {
    const initialState = {
        selectedMovie: {
            data: {
                id: 424785,
                title: "Transformers 7",
                tagline: "",
                vote_average: 0,
                vote_count: 0,
                release_date: "2019-06-26",
                poster_path: "https://image.tmdb.org/t/p/w500/432BowXw7a4fWXSONxBaFKqvW4f.jpg",
                overview: "Plot unknown.",
                budget: 0,
                revenue: 0,
                runtime: null,
            },
            isFetching: false,
            err: null
        }
    };
    const mockStore = configureMockStore();

    let store, container, component;
    
    beforeEach(() => {
        store = mockStore(initialState);
        container = mount(<Provider store={store}><MovieDetail id={123}/></Provider>);
        component = container.find('MovieDetail');
    });
    
    it("renders correctly", () => {
        expect(container).toMatchSnapshot();
        expect(component.prop('selectedMovie')).toBe(initialState.selectedMovie.data);
    });
    
    it("selected movie was reset on unmount event", () => {
        const instance = component.instance();
        instance.componentWillUnmount();
        expect(store.getActions()).toContainEqual(resetMovie());
    });
  });