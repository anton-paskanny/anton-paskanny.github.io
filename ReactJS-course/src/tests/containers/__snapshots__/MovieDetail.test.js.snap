// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetail container renders correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(MovieDetail)
    id={123}
  >
    <MovieDetail
      fetchMovie={[Function]}
      id={123}
      isFetching={false}
      resetMovie={[Function]}
      selectedMovie={
        Object {
          "budget": 0,
          "id": 424785,
          "overview": "Plot unknown.",
          "poster_path": "https://image.tmdb.org/t/p/w500/432BowXw7a4fWXSONxBaFKqvW4f.jpg",
          "release_date": "2019-06-26",
          "revenue": 0,
          "runtime": null,
          "tagline": "",
          "title": "Transformers 7",
          "vote_average": 0,
          "vote_count": 0,
        }
      }
    >
      <div
        className="movie-detail"
      >
        <figure
          className="movie-detail__img-placeholder"
        >
          <img
            alt="Transformers 7"
            className="movie-detail__img"
            src="https://image.tmdb.org/t/p/w500/432BowXw7a4fWXSONxBaFKqvW4f.jpg"
          />
        </figure>
        <div
          className="movie-detail__info-wrapper"
        >
          <div
            className="movie-detail__title-wrapper"
          >
            <h2
              className="movie-detail__title"
            >
              Transformers 7
            </h2>
            <p
              className="movie-detail__rating"
            >
              <span>
                0
              </span>
            </p>
          </div>
          <ul
            className="movie-detail__genres"
          />
          <p
            className="movie-detail__wrapper"
          >
            <time
              className="movie-detail__release-date"
              dateTime="2019-06-26"
            >
              2019
               min
            </time>
            <span
              className="movie-detail__duration"
            />
          </p>
          <p
            className="movie-detail__desc"
          >
            Plot unknown.
          </p>
        </div>
      </div>
    </MovieDetail>
  </Connect(MovieDetail)>
</Provider>
`;
